### ISSUE TITLE
[rush] Performance regression in ProjectChangeAnalyzer

### ISSUE THREAD

**Issue by dmichon-msft** [microsoft/rushstack#3750]
## Summary
Comparing the time spent in "Analyzing repo state" between @microsoft/rush 5.80.0 (version currently installed in the rushstack repo) and @microsoft/rush 5.83.3 (current main) shows over a second of performance regression, which in the repo is 200%.

```diff
-PS D:\rushstack> rush build -o heft-web-rig
+PS D:\rushstack> node --cpu-prof ./apps/rush/lib/start-dev.js build -o heft-web-rig

-Rush Multi-Project Build Tool 5.80.0 - https://rushjs.io
+Rush Multi-Project Build Tool 5.83.3 (unmanaged) - https://rushjs.io
Node.js version is 14.20.0 (LTS)


Starting "rush build"

-Analyzing repo state... DONE (0.56 seconds)
+Analyzing repo state... DONE (1.76 seconds)

Executing a maximum of 7 simultaneous processes...

==[ @rushstack/heft-web-rig (build) ]==============================[ 1 of 1 ]==
"@rushstack/heft-web-rig (build)" did not define any work.



==[ NO OP: 1 operation ]=======================================================

These operations did not define any work:
  @rushstack/heft-web-rig (build)


-rush build (0.68 seconds)
+rush build (1.86 seconds)
```

## Repro steps
Compare the "Analyzing repo state" stage of `node ./apps/rush/lib/start-dev.js build -o heft-web-rig` vs. `rush build -o heft-web-rig`.

## Details
![image](https://user-images.githubusercontent.com/26827560/202013686-b11f3cbe-8511-44ad-a69c-66082cbd90bd.png)
CPU Profiling shows the time is all spent in the `applySubmodulesTreeState` function, despite the rushstack repository not having any submodules.

Related #3749.

Recommended fix would be to make submodules handling an opt-in setting in rush.json, or find an alternate code path to detect that submodules are present, since the check for submodule status takes longer than the entire rest of the process twice over.

Also worth investigating is switching `_getRepoDeps` in `ProjectChangeAnalyzer` to be async, and creating `getRepoStateAsync` that does as many of the Git operations as possible in parallel. The `git status` call and resulting `git hash-object` still have to be serial, but the `git ls-tree` and the `ProjectChangeAnalyzer`'s own `git hash-object` call can be parallelized with the rest to reduce the latency of operations.

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? | 5.82.0 |
| `rushVersion` from rush.json? | 5.80.0 |
| `useWorkspaces` from rush.json? | true |
| Operating system? | Windows |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | 14.20.0 |

**Comment by dmichon-msft** [comment]
Looks like git submodules show up in the `git ls-tree` response with the objecttype `commit` as opposed to `blob`, so we can get an enumeration of them from the existing `git ls-tree` call.
There will also be a `.gitmodules` file if any submodules are present.

**Comment by dmichon-msft** [comment]
Proposed solution:
1) Update `parseGitLsTree` to return an object:
```ts
interface IGitTreeState {
  files: Map<string, string>; // type "blob"
  submodules: Map<string, string>; // type "commit"
}
```
Then use Async.forEachAsync over the list of discovered submodules to run the same `git ls-tree` command again using the submodule as the working directory. Repeat as needed. If the submodules are allowed to be dirty, we will need to also run the `git status` command in the submodules.

**Comment by chengcyber** [comment]
Thanks @dmichon-msft for the investigation on the performance regression from supporting git submodules!

> Have the penultimate output from getRepoState be an IGitTreeState object,

Originally, i've thought about this idea before. The concern was i can not found a doc to describe the fact that object type "commit" points to a `submodule` path. So, `git submodule status` is chosen as the final implementation. 

I am willing to contribute this proposal if it works.

**Comment by iclanton** [comment]
@dmichon-msft also mentioned that we could do a quick existence check for a `.gitmodules` file in the root of the repo if we don't want to trust that object type `"commit"` refers to a submodule.

**Comment by iclanton** [comment]
@chengcyber - It'd be great if you would put together a fix.

**Comment by dmichon-msft** [comment]
I have a draft PR here with some ideas in place that help a lot with the perf of the non-submodule code path: https://github.com/microsoft/rushstack/pull/3756
If we do detection in the `git ls-tree` call it should be possible to adapt that code to work with submodules just fine.

**Comment by chengcyber** [comment]
Hi @dmichon-msft , @iclanton , though David is working on introducing `getRepoStateAsync` in another PR, I created PR https://github.com/microsoft/rushstack/pull/3758 to fix the performance regression for the sync version of `getRepoState`.
